File: .pi/skills/power-analysis/scripts/power.py
#!/usr/bin/env python3
"""Power analysis (analytic) — statsmodels cross-check to power.R.

Supports: two.sample.t, paired.t, anova, correlation, regression.
Given two of {effect, power, n}, solves for the third.

Examples:
  python3 power.py --test two.sample.t --d 0.5 --power 0.8        # n per group
  python3 power.py --test two.sample.t --d 0.5 --n 64             # achieved power
  python3 power.py --test two.sample.t --power 0.8 --n 64        # MDE (d)
  python3 power.py --test anova --f 0.25 --k 3 --power 0.8
  python3 power.py --test correlation --r 0.3 --power 0.8
  python3 power.py --test regression --f2 0.15 --power 0.8
"""
import argparse
import math
import sys


def _have_statsmodels():
    try:
        import statsmodels  # noqa: F401
        return True
    except Exception:
        return False


def _fmt(x, nd=3):
    if isinstance(x, float):
        return f"{x:.{nd}f}"
    return str(x)


def _explicit_arg(name):
    """Return the int value of a CLI arg only if it was explicitly passed on the command line."""
    import sys as _sys
    flag = f"--{name}"
    if flag in _sys.argv:
        i = _sys.argv.index(flag)
        if i + 1 < len(_sys.argv):
            try:
                return int(_sys.argv[i + 1])
            except ValueError:
                return None
    return None


# ---------------------------------------------------------------- main
def main(argv=None):
    p = argparse.ArgumentParser(
        prog="power.py",
        description="Analytic power analysis (statsmodels). Cross-checks power.R. "
                    "Given two of {effect, power, n}, solves for the third.",
    )
    p.add_argument("--test", required=True,
                   choices=["two.sample.t", "paired.t", "anova", "correlation", "regression"],
                   help="Design / test type.")
    p.add_argument("--d", type=float, default=None,
                   help="Cohen's d (for t-tests).")
    p.add_argument("--f", type=float, default=None,
                   help="Cohen's f (for ANOVA).")
    p.add_argument("--f2", type=float, default=None,
                   help="Cohen's f² (for regression).")
    p.add_argument("--r", type=float, default=None,
                   help="Correlation coefficient r (for correlation test).")
    p.add_argument("--k", type=int, default=2,
                   help="Number of groups (ANOVA). Default 2.")
    p.add_argument("--power", type=float, default=None,
                   help="Target power (e.g. 0.80). Omit to solve for power given --n.")
    p.add_argument("--n", type=int, default=None,
                   help="Sample size per group. Omit to solve for n given --power.")
    p.add_argument("--alpha", type=float, default=0.05,
                   help="Significance level α. Default 0.05 (two-sided).")
    p.add_argument("--fmri", action="store_true",
                   help="Append the fMRI cluster/voxel power caveat.")
    args = p.parse_args(argv)

    if not _have_statsmodels():
        print("ERROR: statsmodels is not installed. Install with: pip install statsmodels", file=sys.stderr)
        print("       (CHATLabAI install.sh installs it into the project venv.)", file=sys.stderr)
        return 2

    # Determine which variable to solve for.
    have_effect = any(v is not None for v in (args.d, args.f, args.f2, args.r))
    if not have_effect and args.power is not None and args.n is not None:
        solve = "mde"  # solve for effect size given n + power
    elif have_effect and args.power is not None and args.n is None:
        solve = "n"
    elif have_effect and args.n is not None and args.power is None:
        solve = "power"
    else:
        p.error("Provide exactly two of: effect size, --power, --n. "
                "(e.g. --d 0.5 --power 0.8  =>  n;  --d 0.5 --n 64  =>  power;  "
                "--power 0.8 --n 64  =>  MDE)")

    from statsmodels.stats.power import (TTestIndPower, TTestPower, FTestAnovaPower,
                                          FTestPowerF2, GofChisquarePower)

    alpha = args.alpha

    # ---- t-tests
    if args.test in ("two.sample.t", "paired.t"):
        is_two_sample = args.test == "two.sample.t"
        metric = "Cohen's d"
        eff = args.d
        if eff is None and solve != "mde":
            p.error(f"--d (Cohen's d) is required for {args.test}.")

        # TTestIndPower uses nobs1 + ratio; TTestPower (paired) uses nobs (no ratio).
        def _solve_t(effect_size, power, n):
            if is_two_sample:
                return TTestIndPower().solve_power(
                    effect_size=effect_size, nobs1=n, alpha=alpha,
                    power=power, ratio=1.0, alternative="two-sided")
            return TTestPower().solve_power(
                effect_size=effect_size, nobs=n, alpha=alpha,
                power=power, alternative="two-sided")

        if solve == "n":
            n = _solve_t(effect_size=eff, power=args.power, n=None)
            n_per = int(math.ceil(n))
            _report(args, "n per group", n_per, f"{metric} = {_fmt(eff)}",
                    f"power = {_fmt(args.power)}", f"α = {_fmt(alpha)} (two-sided)",
                    plain=(f"A {args.test} detecting {metric}={_fmt(eff)} at α={_fmt(alpha)} "
                           f"(two-sided) with {_fmt(args.power)} power requires approximately "
                           f"{n_per} participants per group."))
        elif solve == "power":
            power = _solve_t(effect_size=eff, power=None, n=args.n)
            _report(args, "achieved power", power, f"{metric} = {_fmt(eff)}",
                    f"n = {args.n} per group", f"α = {_fmt(alpha)} (two-sided)",
                    plain=(f"With n={args.n} per group, a {args.test} has approximately "
                           f"{_fmt(power)} power to detect {metric}={_fmt(eff)} at α={_fmt(alpha)}."))
        else:  # mde
            eff = _solve_t(effect_size=None, power=args.power, n=args.n)
            _report(args, "minimum detectable d", eff, f"n = {args.n} per group",
                    f"power = {_fmt(args.power)}", f"α = {_fmt(alpha)} (two-sided)",
                    plain=(f"With n={args.n} per group and {_fmt(args.power)} power, the minimum "
                           f"detectable {metric} at α={_fmt(alpha)} is approximately d={_fmt(eff)}."))

    # ---- ANOVA
    elif args.test == "anova":
        power_cls = FTestAnovaPower()
        metric = "Cohen's f"
        eff = args.f
        if eff is None and solve != "mde":
            p.error("--f (Cohen's f) is required for anova.")
        k = args.k
        if solve == "n":
            n = power_cls.solve_power(effect_size=eff, power=args.power, alpha=alpha,
                                     k_groups=k)
            n_per = int(math.ceil(n))
            _report(args, "n per group", n_per, f"{metric} = {_fmt(eff)}",
                    f"k = {k} groups", f"power = {_fmt(args.power)}", f"α = {_fmt(alpha)}",
                    plain=(f"A one-way ANOVA with k={k} groups detecting {metric}={_fmt(eff)} "
                           f"at α={_fmt(alpha)} with {_fmt(args.power)} power requires approximately "
                           f"{n_per} participants per group (N={n_per*k} total)."))
        elif solve == "power":
            power = power_cls.solve_power(effect_size=eff, nobs=args.n, alpha=alpha, k_groups=k)
            _report(args, "achieved power", power, f"{metric} = {_fmt(eff)}",
                    f"n = {args.n} per group", f"k = {k} groups", f"α = {_fmt(alpha)}",
                    plain=(f"With n={args.n} per group (k={k}), a one-way ANOVA has approximately "
                           f"{_fmt(power)} power to detect {metric}={_fmt(eff)} at α={_fmt(alpha)}."))
        else:
            eff = power_cls.solve_power(effect_size=None, power=args.power, nobs=args.n,
                                        alpha=alpha, k_groups=k)
            _report(args, "minimum detectable f", eff, f"n = {args.n} per group",
                    f"k = {k} groups", f"power = {_fmt(args.power)}", f"α = {_fmt(alpha)}",
                    plain=(f"With n={args.n} per group (k={k}) and {_fmt(args.power)} power, the minimum "
                           f"detectable {metric} at α={_fmt(alpha)} is approximately f={_fmt(eff)}."))

    # ---- correlation
    elif args.test == "correlation":
        power_cls = TTestPower()  # test of r uses t-distribution
        metric = "r"
        eff = args.r
        if eff is None and solve != "mde":
            p.error("--r (correlation) is required for correlation.")
        if solve == "n":
            n = power_cls.solve_power(effect_size=eff, power=args.power, alpha=alpha,
                                     alternative="two-sided")
            n_total = int(math.ceil(n))
            _report(args, "n total", n_total, f"{metric} = {_fmt(eff)}",
                    f"power = {_fmt(args.power)}", f"α = {_fmt(alpha)} (two-sided)",
                    plain=(f"A correlation test detecting r={_fmt(eff)} at α={_fmt(alpha)} "
                           f"(two-sided) with {_fmt(args.power)} power requires approximately "
                           f"{n_total} participants."))
        elif solve == "power":
            power = power_cls.solve_power(effect_size=eff, nobs=args.n, alpha=alpha,
                                          alternative="two-sided")
            _report(args, "achieved power", power, f"{metric} = {_fmt(eff)}",
                    f"n = {args.n}", f"α = {_fmt(alpha)} (two-sided)",
                    plain=(f"With n={args.n}, a correlation test has approximately {_fmt(power)} "
                           f"power to detect r={_fmt(eff)} at α={_fmt(alpha)}."))
        else:
            eff = power_cls.solve_power(effect_size=None, power=args.power, nobs=args.n,
                                         alpha=alpha, alternative="two-sided")
            _report(args, "minimum detectable r", eff, f"n = {args.n}",
                    f"power = {_fmt(args.power)}", f"α = {_fmt(alpha)} (two-sided)",
                    plain=(f"With n={args.n} and {_fmt(args.power)} power, the minimum detectable "
                           f"correlation at α={_fmt(alpha)} is approximately r={_fmt(eff)}."))

    # ---- regression
    elif args.test == "regression":
        power_cls = FTestPowerF2()
        metric = "Cohen's f²"
        eff = args.f2
        if eff is None and solve != "mde":
            p.error("--f2 (Cohen's f²) is required for regression.")
        # F-test for regression: numerator df = number of predictors (default 1).
        # --k defaults to 2 (for ANOVA); for regression use 1 unless the user explicitly
        # passes a value > 1. We detect "explicitly passed" via command-line presence.
        k_raw = _explicit_arg("k")
        df_num = k_raw if k_raw is not None and k_raw > 1 else 1
        if solve == "n":
            # Solve for residual df, then n = resid_df + df_num + ncc(=1).
            resid_df = power_cls.solve_power(effect_size=eff, power=args.power,
                                             df_num=df_num, alpha=alpha)
            n_total = int(math.ceil(resid_df)) + df_num + 1
            _report(args, "n total", n_total, f"{metric} = {_fmt(eff)}",
                    f"df_num = {df_num} predictors", f"power = {_fmt(args.power)}", f"α = {_fmt(alpha)}",
                    plain=(f"A multiple regression ({df_num} predictor(s)) detecting {metric}={_fmt(eff)} "
                           f"at α={_fmt(alpha)} with {_fmt(args.power)} power requires approximately "
                           f"{n_total} participants."))
        elif solve == "power":
            resid_df = args.n - df_num - 1
            if resid_df < 1:
                p.error(f"n={args.n} is too small for {df_num} predictor(s) (need n > {df_num + 1}).")
            power = power_cls.solve_power(effect_size=eff, df_num=df_num, df_denom=resid_df,
                                          alpha=alpha)
            _report(args, "achieved power", power, f"{metric} = {_fmt(eff)}",
                    f"n = {args.n}", f"df_num = {df_num} predictors", f"α = {_fmt(alpha)}",
                    plain=(f"With n={args.n}, a regression ({df_num} predictor(s)) has approximately "
                           f"{_fmt(power)} power to detect {metric}={_fmt(eff)} at α={_fmt(alpha)}."))
        else:
            resid_df = args.n - df_num - 1
            if resid_df < 1:
                p.error(f"n={args.n} is too small for {df_num} predictor(s) (need n > {df_num + 1}).")
            eff = power_cls.solve_power(effect_size=None, power=args.power, df_num=df_num,
                                        df_denom=resid_df, alpha=alpha)
            _report(args, "minimum detectable f²", eff, f"n = {args.n}",
                    f"df_num = {df_num} predictors", f"power = {_fmt(args.power)}", f"α = {_fmt(alpha)}",
                    plain=(f"With n={args.n} and {_fmt(args.power)} power, the minimum detectable "
                           f"{metric} at α={_fmt(alpha)} is approximately f²={_fmt(eff)}."))

    if args.fmri:
        print()
        print("fMRI note: cluster/voxel-level power is approximate. This estimate covers the "
              "behavioral-design sample size; for fMRI, state smoothness, cluster-forming "
              "threshold, and effect location explicitly and treat as a planning estimate.")

    return 0


def _report(args, what, value, *assumptions, plain=""):
    print("=" * 60)
    print(f"  Power analysis — {args.test}")
    print("=" * 60)
    print(f"  Result: {what} = {value}")
    print("  Assumptions:")
    for a in assumptions:
        print(f"    - {a}")
    if plain:
        print()
        print(f"  {plain}")
    print("=" * 60)


if __name__ == "__main__":
    sys.exit(main())

File: .pi/skills/power-analysis/scripts/power.R
#!/usr/bin/env Rscript
# Power analysis (analytic + simulation) — pwr + simr/lme4.
# Companion to power.py (cross-check).
#
# Usage:
#   Rscript power.R --test two.sample.t --d 0.5 --power 0.8
#   Rscript power.R --test paired.t       --d 0.5 --power 0.8
#   Rscript power.R --test anova          --f 0.25 --k 3 --power 0.8
#   Rscript power.R --test correlation    --r 0.3 --power 0.8
#   Rscript power.R --test regression     --f2 0.15 --power 0.8
#   Rscript power.R --test two.sample.t --d 0.5 --n 64            # power given n
#   Rscript power.R --test mixed.sim --help                      # simr example (documented)

# ---- minimal argument parser (avoid optparse dependency) -------------------
args <- commandArgs(trailingOnly = TRUE)

get_arg <- function(name, default = NA) {
  idx <- which(args == name)
  if (length(idx) == 0) return(default)
  if (idx == length(args)) return(default)
  args[idx + 1]
}
has_flag <- function(name) name %in% args

print_usage <- function() {
  cat("Usage: Rscript power.R --test <TYPE> [options]\n")
  cat("\nTypes: two.sample.t, paired.t, anova, correlation, regression, mixed.sim\n")
  cat("\nOptions:\n")
  cat("  --d <float>      Cohen's d (t-tests)\n")
  cat("  --f <float>      Cohen's f (ANOVA)\n")
  cat("  --f2 <float>     Cohen's f^2 (regression)\n")
  cat("  --r <float>      correlation r\n")
  cat("  --k <int>        number of groups (ANOVA, default 2)\n")
  cat("  --power <float>  target power (e.g. 0.80)\n")
  cat("  --n <int>        sample size per group\n")
  cat("  --alpha <float>  significance level (default 0.05)\n")
  cat("  --fmri           append fMRI cluster/voxel power caveat\n")
  cat("  --help           show this message\n")
  cat("\nProvide exactly two of: effect size, --power, --n.\n")
  cat("( --d 0.5 --power 0.8 => n;  --d 0.5 --n 64 => power;  --power 0.8 --n 64 => MDE )\n")
}

if (length(args) == 0 || has_flag("--help") || has_flag("-h")) {
  print_usage()
  if (length(args) > 0 && (has_flag("--help") || has_flag("-h")) && "--test" %in% args) {
    test <- get_arg("--test", "")
    if (test == "mixed.sim") {
      cat("\n--- mixed.sim (simr example) ---\n")
      cat("Estimates power for a repeated-measures mixed-effects model via simulation.\n")
      cat("Builds an lme4::lmer model with a fixed effect, sets its effect size, then uses\n")
      cat("simr::powerCurve to estimate power across sample sizes.\n")
      cat("Requires R packages: lme4, simr (and lmerTest optional).\n\n")
      cat("Example model (repeated-measures aesthetics paradigm):\n")
      cat("  rating ~ condition + (1 + condition | subject) + (1 | item)\n")
      cat("  condition effect coded; fixed effect size set via fixef(model)['condition'] <- d\n")
      cat("  powerCurve extends the number of subjects.\n\n")
      cat("To run: Rscript power.R --test mixed.sim --d 0.4 --n 30\n")
    }
  }
  quit(status = 0)
}

test <- get_arg("--test", NA)
if (is.na(test)) {
  cat("ERROR: --test is required. Use --help for usage.\n")
  quit(status = 2)
}

alpha <- as.numeric(get_arg("--alpha", "0.05"))
power_target <- as.numeric(get_arg("--power", NA))
n_given <- as.integer(get_arg("--n", NA))
d <- as.numeric(get_arg("--d", NA))
f <- as.numeric(get_arg("--f", NA))
f2 <- as.numeric(get_arg("--f2", NA))
r <- as.numeric(get_arg("--r", NA))
k <- as.integer(get_arg("--k", "2"))
fmri <- has_flag("--fmri")

report <- function(test, what, value, assumptions, plain) {
  cat(strrep("=", 60), "\n")
  cat(sprintf("  Power analysis (R) - %s\n", test))
  cat(strrep("=", 60), "\n")
  cat(sprintf("  Result: %s = %s\n", what, value))
  cat("  Assumptions:\n")
  for (a in assumptions) cat(sprintf("    - %s\n", a))
  cat("\n")
  cat(sprintf("  %s\n", plain))
  cat(strrep("=", 60), "\n")
  if (fmri) {
    cat("\nfMRI note: cluster/voxel-level power is approximate. State smoothness,\n")
    cat("cluster-forming threshold, and effect location explicitly; treat as planning estimate.\n")
  }
}

# ---- check for pwr (analytic cases) ---------------------------------------
need_pwr <- test %in% c("two.sample.t", "paired.t", "anova", "correlation", "regression")
if (need_pwr && !requireNamespace("pwr", quietly = TRUE)) {
  cat("ERROR: R package 'pwr' is not installed.\n", file = "")
  cat("       Install with:  install.packages(\"pwr\")\n")
  cat("       (CHATLabAI install.sh attempts to install pwr into the R user library.)\n")
  cat("       The Python script power.py covers the same analytic cases as a fallback:\n")
  cat("         python3 power.py --test ", test, sep = "", "\n")
  quit(status = 2)
}

# ---- determine what to solve for ------------------------------------------
solve <- NA
have_effect <- !is.na(d) || !is.na(f) || !is.na(f2) || !is.na(r)
if (have_effect && !is.na(power_target) && is.na(n_given)) solve <- "n"
else if (have_effect && !is.na(n_given) && is.na(power_target)) solve <- "power"
else if (!have_effect && !is.na(power_target) && !is.na(n_given)) solve <- "mde"
else if (test == "mixed.sim") solve <- "sim"
else {
  cat("ERROR: provide exactly two of: effect size, --power, --n.\n")
  cat("       ( --d 0.5 --power 0.8 => n;  --d 0.5 --n 64 => power;  --power 0.8 --n 64 => MDE )\n")
  quit(status = 2)
}

# ---- analytic cases via pwr -----------------------------------------------
if (test %in% c("two.sample.t", "paired.t")) {
  if (is.na(d) && solve != "mde") { cat("ERROR: --d (Cohen's d) required.\n"); quit(status = 2) }
  if (solve == "n") {
    res <- if (test == "two.sample.t") pwr::pwr.t.test(d = d, power = power_target, sig.level = alpha, type = "two.sample")
           else pwr::pwr.t.test(d = d, power = power_target, sig.level = alpha, type = "paired")
    nval <- ceiling(res$n)
    report(test, "n per group", nval,
           c(sprintf("Cohen's d = %.3f", d), sprintf("power = %.3f", power_target),
             sprintf("alpha = %.3f (two-sided)", alpha)),
           sprintf("A %s detecting d=%.3f at alpha=%.3f (two-sided) with %.0f%% power requires approximately %g participants per group.",
                   test, d, alpha, power_target * 100, nval))
  } else if (solve == "power") {
    res <- if (test == "two.sample.t") pwr::pwr.t.test(d = d, n = n_given, sig.level = alpha, type = "two.sample")
           else pwr::pwr.t.test(d = d, n = n_given, sig.level = alpha, type = "paired")
    report(test, "achieved power", round(res$power, 3),
           c(sprintf("Cohen's d = %.3f", d), sprintf("n = %g per group", n_given),
             sprintf("alpha = %.3f (two-sided)", alpha)),
           sprintf("With n=%g per group, a %s has approximately %.1f%% power to detect d=%.3f at alpha=%.3f.",
                   n_given, test, res$power * 100, d, alpha))
  } else {
    res <- if (test == "two.sample.t") pwr::pwr.t.test(power = power_target, n = n_given, sig.level = alpha, type = "two.sample")
           else pwr::pwr.t.test(power = power_target, n = n_given, sig.level = alpha, type = "paired")
    report(test, "minimum detectable d", round(res$d, 3),
           c(sprintf("n = %g per group", n_given), sprintf("power = %.3f", power_target),
             sprintf("alpha = %.3f (two-sided)", alpha)),
           sprintf("With n=%g per group and %.0f%% power, the minimum detectable d at alpha=%.3f is approximately %.3f.",
                   n_given, power_target * 100, alpha, res$d))
  }

} else if (test == "anova") {
  if (is.na(f) && solve != "mde") { cat("ERROR: --f (Cohen's f) required.\n"); quit(status = 2) }
  if (solve == "n") {
    res <- pwr::pwr.anova.test(k = k, f = f, power = power_target, sig.level = alpha)
    nval <- ceiling(res$n)
    report(test, "n per group", nval,
           c(sprintf("Cohen's f = %.3f", f), sprintf("k = %g groups", k),
             sprintf("power = %.3f", power_target), sprintf("alpha = %.3f", alpha)),
           sprintf("A one-way ANOVA (k=%g) detecting f=%.3f at alpha=%.3f with %.0f%% power requires ~%g/group (N=%g).",
                   k, f, alpha, power_target * 100, nval, nval * k))
  } else if (solve == "power") {
    res <- pwr::pwr.anova.test(k = k, f = f, n = n_given, sig.level = alpha)
    report(test, "achieved power", round(res$power, 3),
           c(sprintf("Cohen's f = %.3f", f), sprintf("k = %g groups", k),
             sprintf("n = %g per group", n_given), sprintf("alpha = %.3f", alpha)),
           sprintf("With n=%g/group (k=%g), a one-way ANOVA has ~%.1f%% power to detect f=%.3f at alpha=%.3f.",
                   n_given, k, res$power * 100, f, alpha))
  } else {
    res <- pwr::pwr.anova.test(k = k, power = power_target, n = n_given, sig.level = alpha)
    report(test, "minimum detectable f", round(res$f, 3),
           c(sprintf("k = %g groups", k), sprintf("n = %g per group", n_given),
             sprintf("power = %.3f", power_target), sprintf("alpha = %.3f", alpha)),
           sprintf("With n=%g/group (k=%g) and %.0f%% power, the minimum detectable f at alpha=%.3f is ~%.3f.",
                   n_given, k, power_target * 100, alpha, res$f))
  }

} else if (test == "correlation") {
  if (is.na(r) && solve != "mde") { cat("ERROR: --r required.\n"); quit(status = 2) }
  if (solve == "n") {
    res <- pwr::pwr.r.test(r = r, power = power_target, sig.level = alpha)
    nval <- ceiling(res$n)
    report(test, "n total", nval,
           c(sprintf("r = %.3f", r), sprintf("power = %.3f", power_target),
             sprintf("alpha = %.3f (two-sided)", alpha)),
           sprintf("A correlation test detecting r=%.3f at alpha=%.3f with %.0f%% power requires ~%g participants.",
                   r, alpha, power_target * 100, nval))
  } else if (solve == "power") {
    res <- pwr::pwr.r.test(r = r, n = n_given, sig.level = alpha)
    report(test, "achieved power", round(res$power, 3),
           c(sprintf("r = %.3f", r), sprintf("n = %g", n_given),
             sprintf("alpha = %.3f (two-sided)", alpha)),
           sprintf("With n=%g, a correlation test has ~%.1f%% power to detect r=%.3f at alpha=%.3f.",
                   n_given, res$power * 100, r, alpha))
  } else {
    res <- pwr::pwr.r.test(power = power_target, n = n_given, sig.level = alpha)
    report(test, "minimum detectable r", round(res$r, 3),
           c(sprintf("n = %g", n_given), sprintf("power = %.3f", power_target),
             sprintf("alpha = %.3f (two-sided)", alpha)),
           sprintf("With n=%g and %.0f%% power, the minimum detectable r at alpha=%.3f is ~%.3f.",
                   n_given, power_target * 100, alpha, res$r))
  }

} else if (test == "regression") {
  if (is.na(f2) && solve != "mde") { cat("ERROR: --f2 (Cohen's f^2) required.\n"); quit(status = 2) }
  u <- 1  # numerator df = one predictor
  if (solve == "n") {
    res <- pwr::pwr.f2.test(u = u, f2 = f2, power = power_target, sig.level = alpha)
    nval <- ceiling(res$v) + u + 1
    report(test, "n total", nval,
           c(sprintf("Cohen's f^2 = %.3f", f2), sprintf("u = %g (predictors)", u),
             sprintf("power = %.3f", power_target), sprintf("alpha = %.3f", alpha)),
           sprintf("A regression (1 predictor) detecting f^2=%.3f at alpha=%.3f with %.0f%% power requires ~%g participants.",
                   f2, alpha, power_target * 100, nval))
  } else if (solve == "power") {
    v <- n_given - u - 1
    res <- pwr::pwr.f2.test(u = u, f2 = f2, v = v, sig.level = alpha)
    report(test, "achieved power", round(res$power, 3),
           c(sprintf("Cohen's f^2 = %.3f", f2), sprintf("u = %g", u), sprintf("n = %g", n_given),
             sprintf("alpha = %.3f", alpha)),
           sprintf("With n=%g, a regression (1 predictor) has ~%.1f%% power to detect f^2=%.3f at alpha=%.3f.",
                   n_given, res$power * 100, f2, alpha))
  } else {
    v <- n_given - u - 1
    res <- pwr::pwr.f2.test(u = u, power = power_target, v = v, sig.level = alpha)
    report(test, "minimum detectable f^2", round(res$f2, 3),
           c(sprintf("u = %g", u), sprintf("n = %g", n_given), sprintf("power = %.3f", power_target),
             sprintf("alpha = %.3f", alpha)),
           sprintf("With n=%g and %.0f%% power, the minimum detectable f^2 at alpha=%.3f is ~%.3f.",
                   n_given, power_target * 100, alpha, res$f2))
  }

} else if (test == "mixed.sim") {
  # ---- simulation-based power for a repeated-measures mixed model (simr) ----
  # Requires lme4 + simr. Documented example for repeated-measures aesthetics paradigms
  # (subjects rating stimuli across conditions).
  if (!requireNamespace("lme4", quietly = TRUE) || !requireNamespace("simr", quietly = TRUE)) {
    cat("ERROR: R packages 'lme4' and/or 'simr' are not installed.\n")
    cat("       Install with:  install.packages(c(\"lme4\",\"simr\"))\n")
    cat("       (CHATLabAI install.sh attempts to install these into the R user library.)\n")
    cat("       The analytic cases (two.sample.t, anova, etc.) do not require simr.\n")
    quit(status = 2)
  }
  d_sim <- if (!is.na(d)) d else 0.4
  n_sim <- if (!is.na(n_given)) as.integer(n_given) else 30L

  suppressPackageStartupMessages({
    library(lme4)
    library(simr)
  })

  # Build a pilot mixed-effects model for a repeated-measures design:
  #   rating ~ condition + (1 + condition | subject) + (1 | item)
  # 'condition' is effect-coded (-0.5, 0.5). We set the fixed effect (slope) to d_sim
  # to represent the target effect size, then use simr::powerCurve to extend subjects.
  #
  # This is a synthetic pilot (no data file needed) — appropriate when planning a study
  # before data collection.
  set.seed(42)
  n_subj <- n_sim
  n_item <- 20
  dat <- expand.grid(subject = factor(seq_len(n_subj)),
                     item = factor(seq_len(n_item)),
                     condition = c(-0.5, 0.5))
  dat$rating <- rnorm(nrow(dat), mean = 0, sd = 1)

  model <- lmer(rating ~ condition + (1 + condition | subject) + (1 | item), data = dat)

  # Set the fixed effect to the target effect size (Cohen's d on the effect-coded slope).
  fixef(model)["condition"] <- d_sim

  # Power curve: extend the number of subjects from the pilot n upward.
  pc <- powerCurve(model, along = "subject", breaks = c(n_subj, n_subj + 10, n_subj + 20, n_subj + 30),
                   nsim = 100)

  cat(strrep("=", 60), "\n")
  cat("  Power analysis (R) - mixed.sim (simr)\n")
  cat(strrep("=", 60), "\n")
  cat(sprintf("  Model: rating ~ condition + (1 + condition | subject) + (1 | item)\n"))
  cat(sprintf("  Fixed effect 'condition' set to d = %.3f\n", d_sim))
  cat(sprintf("  Pilot: %g subjects x %g items x 2 conditions\n", n_subj, n_item))
  cat(sprintf("  nsim = 100\n"))
  cat("  Power curve (by number of subjects):\n")
  print(pc)
  cat("\n")
  cat("  NOTE: This is a simulation-based estimate for a repeated-measures mixed model.\n")
  cat("  Increase nsim for tighter estimates. Report the curve, not a single number.\n")
  if (fmri) {
    cat("\nfMRI note: cluster/voxel-level power is approximate. State smoothness,\n")
    cat("cluster-forming threshold, and effect location explicitly.\n")
  }
  cat(strrep("=", 60), "\n")

} else {
  cat("ERROR: unknown --test '", test, "'. Use --help.\n", sep = "")
  quit(status = 2)
}
